Upgrade to latest Smithay

ABeltramo opened the pull request:
This is mainly in order to add support for the [pointer constraints](https://wayland.app/protocols/pointer-constraints-unstable-v1) Wayland protocol so that SDL will eventually pick up the relative mouse movements when going outside the edge of the window.

 This turns out to be a huge refactor..

ABeltramo committed:
checkpoint: ported and fixed some errors, added todos..

ABeltramo commented on wayland-display-core/src/wayland/handlers/compositor.rs:
```diff
@@ -26,6 +28,10 @@ impl CompositorHandler for State {
     }
 
+    fn client_compositor_state<'a>(&self, client: &'a Client) -> &'a CompositorClientState {
+        todo!()
```
This trait was required now, but I'm not sure what the implementation expects here. Should I return an empty `CompositorClientState`?

Drakulix replied:
No, we need to initialize that on client creation, store that in the Client and retrieve it (e.g. anvil: https://github.com/Smithay/smithay/blob/master/anvil/src/shell/mod.rs#L104-L106).

For initialization it should be enough to add a `CompositorClientState` to our `ClientState` struct, as that implements `Default`.

ABeltramo commented on wayland-display-core/src/comp/focus.rs:
```diff
@@ -131,6 +134,69 @@ impl PointerTarget<super::State> for FocusTarget {
+    fn frame(&self, seat: &Seat<State>, data: &mut State) {
+        match self {
+            FocusTarget::Wayland(w) => w.frame(seat, data),
+            FocusTarget::Popup(p) => {todo!()},
```
`p.wl_surface().frame()` expects `frame_number` which I'm not sure how to extract in here..

Drakulix replied:
It is probably defaulting to the wrong `frame`-method, try:
```suggestion
            FocusTarget::Popup(p) => PointerTarget::frame(&p.wl_surface(), seat, data),
```

ABeltramo commented on wayland-display-core/src/comp/input.rs:
```diff
@@ -23,14 +24,14 @@ use std::{
-            .map(|fd| unsafe { OwnedFd::from_raw_fd(fd) })
-            .map_err(|err| err as i32)
+        open(path, OFlags::from_bits_truncate(flags as u32), Mode::empty())
+            .map_err(|err| todo!("Convert Errno to i32?"))
```
This is unfortunately just me being not much confident in Rust, I'm sure it's easily done..

Drakulix replied:
```suggestion
            .map_err(|err| err.raw_os_error())
```

ABeltramo commented on wayland-display-core/src/comp/rendering.rs:
```diff
@@ -1,24 +1,32 @@
-pub const CURSOR_DATA_BYTES: &[u8] = include_bytes!("../../resources/cursor.rgba");
+pub const CURSOR_DATA_BYTES: MemoryBuffer = MemoryBuffer::from_slice(
+    include_bytes!("../../resources/cursor.rgba"),
+    Fourcc::Rgba4444, // TODO: Format???
```
Which format is the cursor? Btw, is this the right way to load it up in the new `MemoryBuffer` struct?

Drakulix replied:
Either `Rgba8888` or `Abgr8888`, I don't fully remember, probably the former given the name, but the different apis use different edianess to express the order...

ABeltramo commented on wayland-display-core/src/comp/rendering.rs:
```diff
@@ -45,38 +53,43 @@ impl State {
-                        1.,
-                    )
+                    CursorImageStatus::Named(cursor_icon) => vec![CursorElement::Memory(
+                        // TODO: icon?
```
I saw that there's now this new `Named` instead of `Default` and there's a `cursor_icon` param that goes along with it; I'm not sure where to plug it..

Drakulix replied:
Named means the application just provided a name of an icon to load from the cursor theme. We don't have that right now, so skipping that is *a solution*.

For a proper one check out: https://github.com/Smithay/smithay/blob/master/anvil/src/cursor.rs

ABeltramo commented on wayland-display-core/src/comp/rendering.rs:
```diff
@@ -91,7 +104,7 @@ impl State {
                     self.video_info.as_ref().unwrap().height() as i32,
                 ),
-            ))
+            ), DrmFourcc::Rgba4444) // TODO: format???
```
Which format should I pick here?

Drakulix replied:
Theoretically whatever is in `self.video_info`.

But OpenGL doesn't give us a bunch of formats and the src doesn't advertise anything but 32-bit Rgbx anyway. But the Fourcc-endianess is reversed and we need the alpha variant so `Abgr8888` is the correct one.

ABeltramo commented on wayland-display-core/src/comp/rendering.rs:
```diff
@@ -107,15 +120,15 @@ impl State {
         self.renderer.unbind().map_err(DTRError::Rendering)?;
-
-        Ok((buffer, damage, render_element_states))
+        todo!("Get damage!");
```
Previously `damage` was returned by the `render_output` method, how can I get that here?

Drakulix replied:
`render_output` returns a `RenderOutputResult` now, which contains both the `damage` and the `render_elements_states` now: https://smithay.github.io/smithay/smithay/backend/renderer/damage/struct.RenderOutputResult.html

It *also* contains a sync-point now, which we would need to wait on before giving a hardware buffer to gstreamer or we might not be done writing to it (or even better give that and the dmabuf to gstreamer, but that api change will take a another few years probably...).
But because we copy to memory anyway, OpenGL should already synchronize for us and we can just ignore it.

ABeltramo commented on wayland-display-core/src/wayland/handlers/relative_pointer.rs:
```diff
@@ -1,5 +1,17 @@
+impl OutputHandler for State {}
+impl PointerConstraintsHandler for State{
+    fn new_constraint(&mut self, surface: &WlSurface, pointer: &PointerHandle<Self>) {
+        todo!()
```
I guess this is what needs to be implemented in order to support the pointer constraints protocol; not sure what is expected in that method..

Drakulix replied:
Store that in the state somewhere and then obey the constraint in the input code.

See the PR that implemented this in anvil: https://github.com/Smithay/smithay/pull/872/commits/d978a2b

Also we need to constuct the `PointerConstraintsState` in `src/comp/mod.rs:init` or we never advertise support for it.

(And it would be nice, if this would be put into a separate file like all the other protocol handlers.)

ABeltramo commented on wayland-display-core/src/wayland/handlers/xdg.rs:
```diff
@@ -78,6 +78,10 @@ impl XdgShellHandler for State {
     }
+
+    fn reposition_request(&mut self, surface: PopupSurface, positioner: PositionerState, token: u32) {
+        todo!()
```
This traits was required now, I'm not sure how to implement it..

Drakulix replied:
We don't properly compute pointer constraints in this compositor at all, so we can't really do this.

The good thing is, this used to be much more complicated, but a lot of helpers were recently implemented properly in smithay. Just copy what smallvil does: https://github.com/Smithay/smithay/pull/1256/files#diff-f471a178782c2c9f3ca35beb0d6d87a219705ba01eaa0f9f1489916f44a7fbcc

Which just makes sure the popup is always visible on the current output (which is exactly what we need for our usecase).

ABeltramo commented on wayland-display-core/src/wayland/protocols/wl_drm.rs:
```diff
@@ -156,29 +156,30 @@ where
-                                drm.post_error(wl_drm::Error::InvalidName, "buffer import failed");
-                            }
-                        }
+                        todo!("How can I make a ImportNotifier? Constructor is private..");
```
I've changed  `DmabufHandler::dmabuf_imported` so that it takes an `ImportNotifier` in input instead of returning the result like it used to; it seems that I can't find a way to create it though; the constructor is private, what am I missing?

Drakulix replied:
Yeah, sadly this is the case, so we can't piggy back on the existing `DmabufHandler` implementation any more.

So we either need our own trait and implement that as well (e.g. see how cosmic-comp does this now: https://github.com/pop-os/cosmic-comp/blob/master/src/wayland/protocols/drm.rs#L47) or just drop the `D` and the split between `protocols` and `handler` and just hard-code everything for our state struct, so that we can just call `state.renderer.import_dmabuf` directly here.

ABeltramo reviewed (commented):
I've just done a first pass of things that were simply broken and added a few `todo` here and there to try and get along. The code is still not compiling, so this is far from over but I wanted to start getting some feedback early instead of deep diving further.
@Drakulix when you got time, please send some help 😅

Drakulix commented on wayland-display-core/src/comp/input.rs:
```diff
@@ -60,7 +61,7 @@ impl State {
                                 && !modifiers.logo
                             {
-                                match handle.modified_sym() {
+                                match u32::from(handle.modified_sym()) {
```
Would be cleaner to not change this line and match on the new `Keysym` type instead: https://smithay.github.io/smithay/smithay/input/keyboard/struct.Keysym.html

Drakulix commented on wayland-display-core/src/comp/input.rs:
```diff
@@ -60,7 +61,7 @@ impl State {
                             {
-                                match handle.modified_sym() {
+                                match u32::from(handle.modified_sym()) {
                                     keysyms::KEY_Tab => {
```
So this would become 
```suggestion
                                    Keysym::Tab => {
```

and so on...

Drakulix commented on wayland-display-core/src/comp/input.rs:
```diff
@@ -94,7 +95,7 @@ impl State {
                             }
                         } else {
-                            if data.surpressed_keys.remove(&handle.modified_sym()) {
+                            if data.surpressed_keys.remove(&(u32::from(handle.modified_sym()))) {
```
Little cleaner:
```suggestion
                            if data.surpressed_keys.remove(&handle.modified_sym().raw()) {
```

Drakulix commented on wayland-display-core/src/comp/input.rs:
```diff
@@ -192,29 +193,29 @@ impl State {
                 let horizontal_amount = event
                     .amount(Axis::Horizontal)
-                    .or_else(|| event.amount_discrete(Axis::Horizontal).map(|x| x * 2.0))
+                    .or_else(|| event.amount_v120(Axis::Horizontal).map(|x| x * 2.0))
```
```suggestion
                    .or_else(|| event.amount_v120(Axis::Horizontal).map(|x| x * 2.0 / 120.0))
```

The new v120 are different.

Drakulix commented on wayland-display-core/src/comp/input.rs:
```diff
@@ -192,29 +193,29 @@ impl State {
                 let vertical_amount = event
                     .amount(Axis::Vertical)
-                    .or_else(|| event.amount_discrete(Axis::Vertical).map(|y| y * 2.0))
+                    .or_else(|| event.amount_v120(Axis::Vertical).map(|y| y * 2.0))
```
```suggestion
                    .or_else(|| event.amount_v120(Axis::Vertical).map(|y| y * 2.0 / 120.0))
```

Drakulix reviewed (commented):
Looks pretty good already!

ABeltramo committed:
fix: more errors

ABeltramo committed:
feat: implemented pointer constraint protocol

ABeltramo committed:
fix: implemented unconstrain_popup

ABeltramo committed:
fix: input suggestions

ABeltramo committed:
lost: DRM?

ABeltramo commented on wayland-display-core/src/wayland/handlers/wl_drm.rs:
```diff
@@ -1,3 +1,24 @@
+};
+use smithay::backend::renderer::ImportDma;
+
+impl<D: 'static> DrmHandler<Option<D>> for State {
+    fn dmabuf_imported(
+        &mut self,
+        global: &DmabufGlobal,
+        dmabuf: Dmabuf,
+    ) -> Result<Option<D>, ImportError> {
+        todo!("self.renderer.import_dmabuf(&dmabuf, None) ???");
+        Err(ImportError::Failed)
+    }
+
+    fn buffer_created(&mut self, buffer: WlBuffer, result: Option<D>) {
+        todo!()
+    }
```
I've added the trait `DrmHandler` but I'm a bit lost on how to actually implement this. I was trying to 
```rust
self.renderer.import_dmabuf(&dmabuf, None)
```
but I'm not sure what should I return here.. I've tried `Option<DrmNode>` at first but then it was complaining that
> the trait bound `comp::State: DrmHandler<comp::State>` is not satisfied

I like this approach because everything else in the wl_drm implementation stay the same..

Drakulix replied:
Oh right, the returned value isn't relevant for this compositor (we don't deal with multiple gpus). So just replace the `Option<D>` with `()` (that is the unit type, aka an empty tuple. Used in place of "Nothing").

`buffer_created` can be an empty implementation and `self.renderer.import_dmabuf(&dmabuf, None).map(|_| ()).map_err(|_| ImportError::Failed)` should work for the `dmabuf_imported` method.

ABeltramo commented on wayland-display-core/src/comp/mod.rs:
```diff
@@ -331,16 +342,16 @@ pub(crate) fn init(
                                     .data_map
                                     .get::<XdgToplevelSurfaceData>()
-                                    .map(|attrs| attrs.lock().unwrap().max_size)
+                                    .map(|attrs| todo!("Used to be attrs.lock().unwrap().max_size but max_size is now not present!"))
```
`max_size` is now missing, what am I missing here?

Drakulix replied:
```
states.cached_state.current::<SurfaceCachedState>().max_size
```

ABeltramo commented on wayland-display-core/src/comp/mod.rs:
```diff
@@ -476,27 +487,28 @@ pub(crate) fn init(
-            },
-        )
-        .expect("Failed to init wayland server source");
+    let display_fd = todo!("display.backend().poll_fd().as_raw_fd() -> the trait `AsFd` is not implemented for `i32`");
```
I've tried a few variations using `as_fd()` but it errors out

Drakulix replied:
Just staff the whole display into `Generic::new`. Also I don't think we need the `Data` struct anymore and can just use `State` directly. We can call `flush_clients` now on a `DisplayHandle`, which we already have in `State`.

ABeltramo commented on wayland-display-core/src/comp/focus.rs:
```diff
@@ -117,21 +120,84 @@ impl PointerTarget<super::State> for FocusTarget {
+    fn frame(&self, seat: &Seat<State>, data: &mut State) {
+        match self {
+            FocusTarget::Wayland(w) => w.frame(seat, data),
+            FocusTarget::Popup(p) => { todo!("the trait `PointerTarget<_>` is not implemented for `&WlSurface`") }
```
This is still not working..

Drakulix replied:
Well there certainly is an implementation for it: https://smithay.github.io/smithay/smithay/input/pointer/trait.PointerTarget.html#impl-PointerTarget%3CD%3E-for-WlSurface

How exactly was your attempt looking? Something like this?
```suggestion
            FocusTarget::Popup(p) => { PointerTarget::frame(p.wl_surface(), seat, data) }
```

ABeltramo committed:
fix: FB format

Drakulix commented on wayland-display-core/src/wayland/handlers/compositor.rs:
```diff
@@ -129,7 +128,7 @@ impl CompositorHandler for State {
 
         if let Some(popup) = self.popups.find_popup(surface) {
-            let PopupKind::Xdg(ref popup) = popup;
+            let PopupKind::Xdg(ref popup) = popup else { todo!("refutable pattern in local binding") };
```
Ah right, we also now have input method handling in the popup code. Our compositor doesn't do that, so you can just replace the `todo!` with `unreachable!()`, so it will (most likely) get optimized away.

ABeltramo committed:
fix: removed extra Data struct, use State directly in event loop

ABeltramo committed:
fix: DrmHandler implementation and other missing bits

ABeltramo committed:
fix: formats from video info, added sync before sending gstreamer buffer

ABeltramo commented:
The good news is that the compiler is happy and I can finally build and test this; the bad news is that there's nothing in the stream..

The unit test in Wolf where I'm simply importing the buffer and checking it is failing, but when I debug it, it works. So I'm wondering if it's related to your comment about the sync.. I've added this but it might not be enough or not in the right spot?
https://github.com/games-on-whales/gst-wayland-display/blob/63d9e80081d54293718a26590e4ff8c19513b1b5/wayland-display-core/src/comp/mod.rs#L386-L391

Also, I've added a call to go from `info.format()` instead of hardcoding it
https://github.com/games-on-whales/gst-wayland-display/blob/63d9e80081d54293718a26590e4ff8c19513b1b5/wayland-display-core/src/comp/mod.rs#L334-L338
but I'm not sure it's working properly, 
https://github.com/games-on-whales/gst-wayland-display/blob/63d9e80081d54293718a26590e4ff8c19513b1b5/wayland-display-core/src/comp/mod.rs#L300
prints out 
> Requested video format: RGBx .to_fourcc() = 0
so I guess it's going to the hardcoded path anyway.

I'll debug further what's going on with the `GstBuffer` on Wolf side.

ABeltramo commented:
Or maybe simply this missing bit should be implemented..
https://github.com/games-on-whales/gst-wayland-display/blob/63d9e80081d54293718a26590e4ff8c19513b1b5/wayland-display-core/src/comp/mod.rs#L499-L514

ABeltramo committed:
fix: missing dispatch_clients

ABeltramo committed:
fix: added missing PointerConstraintsState

ABeltramo committed:
fix: warnings about unused requirements

ABeltramo committed:
fix: mouse scroll not working
